src/FdoActionGroup.vala

```diff
@@ -0,0 +1,214 @@
+                    out state_type,
+                    out state_hint,
+                    out state
+                );
+            } else {
+                enabled = false;
+            }
+        }
+
+        return enabled;
+    }
+
+    public string[] list_actions () {
+        if (actions_list != null) {
+            return actions_list.data;
+        }
+
+        actions_list = new GenericArray<string> ();
+        actions_list.add ("close");
+
+        foreach (var group in groups.get_values ()) {
+            actions_list.extend (group.actions, (action) => ACTION_ID.printf (group.id, action));
+        }
+
+        return actions_list.data;
+    }
+
+    public void activate_action (string action, Variant? target)
+    requires (action in list_actions ()) {
+        if (action == "close" && (uint32) target in groups) {
+            server.notification_closed ((uint32) target, Notifications.Server.CloseReason.DISMISSED);
+        } else {
+            long name_offset = 0;
+            uint32 id = 0;
+
+            if (action.scanf (ACTION_SCANF, ref id, ref name_offset) > 0) {
+                var action_name = action.substring (name_offset);
+                debug ("activating action '%s' for notification id '%u'", action_name, id);
+                groups[id].activate_action (action_name, target);
+            } else {
+                warning ("failed to activate action '%s': failed to parse notification id", action);
+            }
+        }
+    }
+
+    /* XXX: GLib says that we are only meant to override list_actions and query_actions, however
+     *      the gio bindings force us implement everything.
+     */
+    public void change_action_state (string action_name, Variant @value) {
+    }
+
+    public bool get_action_enabled (string action_name) {
+        return has_action (action_name);
+    }
+
+    public unowned VariantType? get_action_parameter_type (string action_name) {
+        return action_name == "close" ? VariantType.UINT32 : null;
+    }
+
+    public Variant? get_action_state (string action_name) {
+        return null;
+    }
+
+    public Variant? get_action_state_hint (string action_name) {
+        return null;
+    }
+
+    public unowned VariantType? get_action_state_type (string action_name) {
+        return null;
+    }
+
+    public bool has_action (string action_name) {
+        return action_name == "close" || action_name in list_actions ();
+    }
+}
+
+private class Notifications.FdoNotificationActionGroup : Object, ActionGroup {
+    public uint id { get; set; }
+    public GenericArray<string> actions { get; set; }
+
+    public FdoNotificationActionGroup (uint id, string[] actions_name) {
+        // TODO: Use GenericArray.null_terminated() when glib_version >= 2.74
+        Object (id: id, actions: new GenericArray<string> (actions_name.length + 1));
+        foreach (var action_name in actions_name) {
+            actions.add (action_name);
+        }
+
+        actions.add (null);
```

Comment:
This currently gives a compile time warning and is probably also responsible for this run time critical: `____lambda12_: assertion 'action != NULL' failed`. Also I think (because in my testing with this removed it worked) it causes the `list_actions` method of the `FdoActionGroup` to not work which in turn causes the action activation on the main group e.g. via dbus to not work.
Maybe for now don't add it or would that be problematic?

Reply:
Yes it's for both, however there's no other way to make a GenericArray null-terminated with our current version of GLib.

It's funny because i added that null exactly because `list_actions()` was causing segfaults during my tests.

i can activate actions via DBus here still. (remember that the server expire the notification when the bubble timeouts, that is not resolved here).

Reply:
Maybe we can simply use a `string[]` type here? You can now easily build them with `GLib.StrvBuilder`

Reply:
thanks for the tip! didn't known about StrvBuilder.